### PR TITLE
[PULL REQUEST] Add sulfur in-cloud and on seasalt reactions via KPP (Supersedes #934)

### PR DESCRIPTION
**Pull request by yantosca**
This PR is a continuation of #934.  We have merged the latest updates in implemeting sulfur in-cloud and on seasalt reactions via KPP on top of 13.4.0-alpha.5, so that we could take advantage of recent bug fixes for heterogeneous chemistry.

Before merging, we will examine internal benchmark output w/r/t 13.3.0 and 13.4.0-alpha.5, as well as running the customary integration tests.

### CODE REVIEW

**Comment by yantosca** [comment]
Also tagging @viral211 and @msl3v, feel free to add comments on this PR.

File: KPP/fullchem/fullchem_RateLawFuncs.F90 [hunk 1]
```diff
@@ -2078,82 +2074,10 @@ FUNCTION HOBrHetSO4H2( H ) RESULT( k )
...
-!       !
-!    ENDIF
-!
-!    ! Assume HOBr is limiting, so update the removal rate accordingly
-!    k = kIIR1Ltd( C(ind_HOBr), C(ind_SO3mm), k )
-!  END FUNCTION HOBrUptkBySO3mm
-!-------------------------------------------------------------------------------
+
+    ! Assume HOBr is limiting, so update the removal rate accordingly
+    k = kIIR1Ltd( C(ind_HOBr), C(ind_SO3mm), k )
```
Comment (laestrada) [inline KPP/fullchem/fullchem_RateLawFuncs.F90:2079]: These are some wild function names

File: KPP/fullchem/fullchem.eqn [hunk 2]
```diff
@@ -477,31 +474,28 @@ O2         = IGNORE; {O2; Molecular oxygen}
...
 //
-//// Seasalt
-//SO2  + SALAAL + O3  = SO4 - SALAAL  : K_MT(1);
-//HCl  + SALAAL       = SALACL        : K_MT(2);
-//HNO3 + SALAAL       = NIT           : K_MT(3);
-//SO2  + SALCAL + O3  = SO4s - SALCAL : K_MT(4);
-//HCl  + SALCAL       = SALCCL        : K_MT(5);
-//HNO3 + SALCAL       = NITs          : K_MT(6);
+// Seasalt
+SO2  + SALAAL + O3  = SO4 - SALAAL  : K_MT(1);
```
Comment (viral211) [inline KPP/fullchem/fullchem.eqn:482]: Clever!

File: KPP/fullchem/fullchem.eqn [hunk 3]
```diff
@@ -1516,16 +1510,16 @@ HOCl + HCl = Cl2 + H2O :       HOClUptkByHCl( State_Het );
...
 HOBr + SALACL = BrCl + H2O :   HOBrUptkBySALACL( State_Het );                                               {2018/01/22; Sherwen2017;TS,JAS,SDE;XW}
 HOBr + SALCCL = BrCl + H2O :   HOBrUptkBySALCCL( State_Het );                                               {2018/01/22; Sherwen2017;TS,JAS,SDE,XW}
 HOBr + BrSALA = Br2 :          HOBrUptkByBrSALA( State_Het );                                               {2017/09/22; Sherwen2017;TS,JAS,SDE}
 HOBr + BrSALC = Br2 :          HOBrUptkByBrSALC( State_Het );                                               {2017/09/22; Sherwen2017;TS,JAS,SDE}
-HOBr = SO4H1 + HBr :           HOBrHetSO4H1( State_Het );                                                   {2017/11/15; Chen2017; QJC}
-HOBr = SO4H2 + HBr :           HOBrHetSO4H2( State_Het );                                                   {2017/11/15; Chen2017; QJC}
+HOBr + HSO3m = SO4s + HBr :     HOBrUptkByHSO3m( State_Het );                                                {2017/11/15; Chen2017; QJC; June 6, 2021, MSL}
+HOBr + SO3mm = SO4s + HBr :     HOBrUptkBySO3mm( State_Het );                                                {2017/11/15; Chen2017; QJC; June 6, 2021, MSL}
```
Comment (viral211) [inline KPP/fullchem/fullchem.eqn:1522]: Should the product be SO4 to replicate the old code? Also for HOCl reactions.

File: KPP/fullchem/fullchem_RateLawFuncs.F90 (hunk 1 above)
Comment (yantosca) [reply KPP/fullchem/fullchem_RateLawFuncs.F90:2079]: I know, a lot of those were in the original code, which I didn't touch.

File: KPP/fullchem/fullchem.eqn (hunk 3 above)
Comment (yantosca) [reply KPP/fullchem/fullchem.eqn:1522]: Not sure, that's what @msl3v (aka @michael-s-long) implemented.  I would let him reply.

Comment (yantosca) [reply KPP/fullchem/fullchem.eqn:1522]: @viral211: For clarification
```
HOBr = SO4H1 + HBr :                         HET(ind_HOBr,7);                 
HOBr = SO4H2 + HBr :                         HET(ind_HOBr,8);   
```
were computed by this old code:
```F90
      !----------------------------------------------------------------------
      ! HOBr + HSO3-(aq) (update: XW 2019-06-08)
      !----------------------------------------------------------------------

      ! This reaction is first order, so no kII calculation is required
      kITemp = HETHOBr_TCld( XDenA,       rLiq,        rIce,                 &
                             ALiq,        AIce,        VAir,                 &
                             TempK,       CldFr,       hConc_Sul,            &
                             hConc_LCl,   hConc_ICl,   clConc_CldA,          &
                             clConc_CldC, clConc_Cldg, brConc_CldA,          &
                             brConc_CldC, brConc_Cldg, HSO3conc_Cld,         &
                             SO3conc_Cld, hno3_th,     hcl_th,               &
                             hbr_th,      7,           H                    )

      ! Make sure sulfate produced is less than SO2 available (qjc, 06/20/16)
      HET(ind_HOBr,  7) = kITemp * fupdateHOBr

      !----------------------------------------------------------------------
      ! HOBr + SO3--(aq) (update: XW 2019-06-08)
      !----------------------------------------------------------------------
      ! This reaction is first order, so no kII calculation is required
      kITemp = HETHOBr_TCld( XDenA,       rLiq,        rIce,                 &
                             ALiq,        AIce,        VAir,                 &
                             TempK,       CldFr,       hConc_Sul,            &
                             hConc_LCl,   hConc_ICl,   clConc_CldA,          &
                             clConc_CldC, clConc_Cldg, brConc_CldA,          &
                             brConc_CldC, brConc_Cldg, HSO3conc_Cld,         &
                             SO3conc_Cld, hno3_th,     hcl_th,               &
                             hbr_th,      8,           H                    )

      ! Make sure sulfate produced is less than SO2 available (qjc, 06/20/16)
      HET(ind_HOBr,  8) = kITemp * fupdateHOBr
```
and similarly for HOCl.

Comment (msl3v) [reply KPP/fullchem/fullchem.eqn:1522]: Viral, I think you're correct. There was (is still, maybe) a drop in SO4s in the new version. And I somehow incorrectly interpreted that as being due to the removal of SO4H[1-4]. This should probably be S4, not SO4s. The question remains: Where did the SO4s go? All I can tell is that SO4s is created in a partitioningoperatoin within CHEM_SO4(). Any thoughts?

Comment (msl3v) [reply KPP/fullchem/fullchem.eqn:1522]: Ah! It looks like SO4s is set from PSO4F from SEASALT_CHEM() and SEASALT_CHEM() isn't being run. I'll try to get a fix in.

**Comment by yantosca** [comment]
FYI, I did some benchmarks (which will probably have to be redone to fix the issue mentioned above), but for what it's worth:

  - [Shows effect of S(IV) -> S(VI) only (w/o HMS)](http://ftp.as.harvard.edu/gcgrid/geos-chem/validation/InternalBenchmarks/13.4.0/GCC_13.4.0-alpha.5/)
  - 
  - [Shows effect of S(IV) -> S(VI) w/ HMS via KPP](http://ftp.as.harvard.edu/gcgrid/geos-chem/validation/InternalBenchmarks/13.4.0/GCC_13.4.0-alpha.7.vs.alpha.5/)

File: KPP/fullchem/fullchem.eqn (hunk 3 above)
Comment (yantosca) [reply KPP/fullchem/fullchem.eqn:1522]: @viral211, @msl3v, maybe this is just the way it is with the KPP implementation.

**Comment by yantosca** [comment]
Added a bug fix commit (c419c25e6861bfbfc6249e93f3759da955bf3cac) to remove from the default GCHP SpeciesConc collection:

  - SpeciesConc_SO4H1
  - SpeciesConc_SO4H2
  - SpeciesConc_SO4H3
  - SpeciesConc_SO4H4

and to add

  - SpeciesConc_HSO3m
  - SpeciesConc_SO3mm

This update had been made before but might have been clobbered when merging the sulfur code via KPP up to the 13.4.0 development branch.

**Comment by yantosca** [comment]
## Update 13 Jan 2020

We have added the following commits to the 13.4.0 development branch:

  1. c7108c3f5 (HEAD -> dev, origin/dev) Set HMS reaction rates K_CLD(4:6) = 0 while we further vet HMS chemistry
  2. 36ec887fa Merge branch 'feature/sulfur-via-kpp' into dev
  3. 94b196f18 (origin/feature/sulfur-via-kpp, feature/sulfur-via-kpp) Merge HMS c

These commits now implement the following sulfur reactions via KPP:
```
// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
// %%%%% Reactions extracted from sulfate_mod.F90 (MSL, BMY)             %%%%%
// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
//
// Seasalt
SO2  + SALAAL + O3  = SO4 - SALAAL       : K_MT(1);
HCl  + SALAAL       = SALACL             : K_MT(2);
HNO3 + SALAAL       = NIT                : K_MT(3);
SO2  + SALCAL + O3  = SO4s - SALCAL      : K_MT(4);
HCl  + SALCAL       = SALCCL             : K_MT(5);
HNO3 + SALCAL       = NITs               : K_MT(6);
//
// Cloud
// S(IV) --> S(VI)
SO2 + H2O2          = SO4                : K_CLD(1);
SO2 + O3            = SO4                : K_CLD(2);
SO2 {+O2}           = SO4                : K_CLD(3);  {Mn & Fe catalysis + HET_DROP_CHEM()}
//
// HMS
CH2O + SO2          = HMS                : K_CLD(4);  {Sep 2021; Moch2020; MSL}
HMS                 = SO2 + CH2O         : K_CLD(5);  {Sep 2021; Moch2020; MSL}
HMS + OH            = 2SO4 + CH2O - SO2  : K_CLD(6);  {Sep 2021; Moch2020; MSL}
```
But the HMS reaction rates `K_CLD(4:6)` are manually set to zero temporarily while we conduct further vetting of the HMS chemistry implementation via KPP.

**Comment by yantosca** [comment]
GEOS-Chem Classic integration tests yield the following results:

```console
==============================================================================
GEOS-Chem Classic: Execution Test Results

Using 24 OpenMP threads
Number of execution tests: 48
==============================================================================
 
Execution tests:
------------------------------------------------------------------------------
gc_025x03125_CH4_geosfp_47L_na...................Execute Simulation.....PASS
gc_025x03125_fullchem_geosfp_47L_na..............Execute Simulation.....PASS
gc_05x0625_CH4_merra2_47L_na.....................Execute Simulation.....PASS
gc_05x0625_fullchem_merra2_47L_na................Execute Simulation.....PASS
gc_2x25_aerosol_merra2...........................Execute Simulation.....PASS
gc_2x25_CH4_merra2...............................Execute Simulation.....PASS
gc_2x25_CO2_merra2...............................Execute Simulation.....PASS
gc_2x25_fullchem_merra2..........................Execute Simulation.....PASS
gc_2x25_Hg_merra2................................Execute Simulation.....PASS
gc_2x25_metals_merra2............................Execute Simulation.....PASS
gc_2x25_POPs_BaP_merra2..........................Execute Simulation.....PASS
gc_2x25_tagCH4_merra2............................Execute Simulation.....PASS
gc_2x25_tagCO_merra2.............................Execute Simulation.....PASS
gc_2x25_tagO3_merra2.............................Execute Simulation.....PASS
gc_2x25_TransportTracers_merra2..................Execute Simulation.....PASS
gc_2x25_TransportTracers_merra2_LuoWd............Execute Simulation.....PASS
gc_4x5_aerosol_geosfp............................Execute Simulation.....PASS
gc_4x5_aerosol_merra2............................Execute Simulation.....PASS
gc_4x5_CH4_geosfp................................Execute Simulation.....PASS
gc_4x5_CH4_merra2................................Execute Simulation.....PASS
gc_4x5_fullchem_aciduptake_merra2................Execute Simulation.....PASS
gc_4x5_fullchem_APM_merra2.......................Execute Simulation.....FAIL
gc_4x5_fullchem_benchmark_merra2.................Execute Simulation.....PASS
gc_4x5_fullchem_complexSOA_merra2................Execute Simulation.....PASS
gc_4x5_fullchem_complexSOA_SVPOA_merra2..........Execute Simulation.....PASS
gc_4x5_fullchem_geosfp...........................Execute Simulation.....PASS
gc_4x5_fullchem_marinePOA_merra2.................Execute Simulation.....PASS
gc_4x5_fullchem_merra2...........................Execute Simulation.....PASS
gc_4x5_fullchem_merra2_47L.......................Execute Simulation.....PASS
gc_4x5_fullchem_merra2_LuoWd.....................Execute Simulation.....PASS
gc_4x5_fullchem_RRTMG_merra2.....................Execute Simulation.....PASS
gc_4x5_fullchem_TOMAS15_merra2_47L...............Execute Simulation.....FAIL
gc_4x5_fullchem_TOMAS40_merra2_47L...............Execute Simulation.....FAIL
gc_4x5_Hg_geosfp.................................Execute Simulation.....PASS
gc_4x5_Hg_merra2.................................Execute Simulation.....PASS
gc_4x5_metals_merra2.............................Execute Simulation.....PASS
gc_4x5_POPs_BaP_geosfp...........................Execute Simulation.....PASS
gc_4x5_POPs_BaP_merra2...........................Execute Simulation.....PASS
gc_4x5_tagCH4_geosfp.............................Execute Simulation.....PASS
gc_4x5_tagCH4_merra2.............................Execute Simulation.....PASS
gc_4x5_tagCO_geosfp..............................Execute Simulation.....PASS
gc_4x5_tagCO_merra2..............................Execute Simulation.....PASS
gc_4x5_tagO3_geosfp..............................Execute Simulation.....PASS
gc_4x5_tagO3_merra2..............................Execute Simulation.....PASS
gc_4x5_TransportTracers_geosfp...................Execute Simulation.....PASS
gc_4x5_TransportTracers_geosfp_LuoWd.............Execute Simulation.....PASS
gc_4x5_TransportTracers_merra2...................Execute Simulation.....PASS
gc_4x5_TransportTracers_merra2_LuoWd.............Execute Simulation.....PASS
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 45
Execution tests failed: 3
Execution tests not yet completed: 0
```
The TOMAS integration tests failed because the default restart files did not have HSO3m, SO3mm.  I am fixing that issue now.

The APM integration test failed at line 1375 of apm_phys_mod.F90.  It might be because I was running on huce_intel instead of huce_cascade on Cannon, and thus less memory was available to the job.

**Comment by yantosca** [comment]
NOTE: Integration tests for GCHP_13.4.0-alpha.14 failed due to a debug print statement (that was used with GEOS-Chem Classic) causing an out of-bounds error.  We have fixed the issue, which has been tagged as 13.4.0-alpha.15.

ALSO NOTE: AWS cloud 1-month benchmarks require a unique tag name.  For this reason we have incremented the tag from 13.4.0-alpha.14 to alpha.15.

**Comment by yantosca** [comment]
We have determined that 13.4.0-alpha.14 had HMS disabled in sulfate_mod.F90.  We have restored the HMS chemistry via KPP in 13.4.0-alpha.17.  

The GEOS-Chem Classic integration tests of 13.4.0-alpha.17 revealed 2 issues in GEOS-Chem Classic APM and TOMAS simulations:
```console
gc_4x5_fullchem_APM_merra2.......................Execute Simulation.....FAIL
gc_4x5_fullchem_TOMAS40_merra2_47L...............Execute Simulation.....FAIL
 
Summary of test results:
------------------------------------------------------------------------------
Execution tests passed: 46
Execution tests failed: 2
Execution tests not yet completed: 0
```
The APM issue has been resolved as described in #1100.  The TOMAS issue is described in #1101.

**Pull request merged**